### PR TITLE
Implement a factory pattern for transition kernels

### PR DESCRIPTION
**Pull request by dmcdougall**
Allows the user to implement a custom transition kernel and register with the factory so QUESO can instantiate the custom kernel at run-time.

All commits pass `make check` locally, the changelog has been updated and documentation included.

This should replace #404 and #408.

### CODE REVIEW

File: src/stats/src/MetropolisHastingsSG.C [hunk 1]
```diff
@@ -514,7 +514,7 @@ MetropolisHastingsSG<P_V,P_M>::commonConstructor()
     msg = "The doLogitTransform option is deprecated.  ";
     msg += "Use ip_mh_algorithm instead.";
     std::cerr << msg << std::endl;
-    queso_warning(msg.c_str());
+    queso_do_once(queso_warning(msg.c_str()));
```
Comment (roystgnr) [inline src/stats/src/MetropolisHastingsSG.C]: queso_warning already is defined to use queso_do_once

File: src/stats/src/MetropolisHastingsSG.C [hunk 2]
```diff
@@ -513,7 +513,6 @@ MetropolisHastingsSG<P_V,P_M>::commonConstructor()
     std::string msg;
     msg = "The doLogitTransform option is deprecated.  ";
     msg += "Use ip_mh_algorithm instead.";
-    std::cerr << msg << std::endl;
```
Comment (roystgnr) [inline src/stats/src/MetropolisHastingsSG.C]: How were you *not* able to replicate this locally?

File: src/stats/src/MetropolisHastingsSG.C (hunk 1 above)
Comment (dmcdougall) [inline src/stats/src/MetropolisHastingsSG.C]: Move along.  Nothing to see here.

File: src/stats/src/MetropolisHastingsSG.C (hunk 2 above)
Comment (dmcdougall) [inline src/stats/src/MetropolisHastingsSG.C]: Move along. Nothing to see here.

**Comment by dmcdougall** [comment]
Ok we're cooking on gas now.

**Pull request merged**